bin/varnishd/cache/cache_range.c

```diff
@@ -180,13 +181,67 @@ vrg_dorange(struct req *req, const char *r, void **priv)
 	return (NULL);
 }
 
+/*
+ * return 1 if range should be observed, based on if-range value
+ * if-range can either be a date or an ETag [RFC7233 3.2 p8]
+ */
+static int
+vrg_ifrange(struct req *req)
+{
+	const char *p, *e;
+	vtim_real ims, lm, d;
+
+	if (!http_GetHdr(req->http, H_If_Range, &p))	// rfc7233,l,455,456
+		return (1);
+
+	/* strong validation needed */
+	if (p[0] == 'W' && p[1] == '/')			// rfc7233,l,500,501
+		return (0);
+
+	/* ETag */
+	if (p[0] == '"') {				// rfc7233,l,512,514
+		if (!http_GetHdr(req->resp, H_ETag, &e))
+			return (0);
+		if ((e[0] == 'W' && e[1] == '/'))	// rfc7232,l,547,548
+			return (0);
+		return (strcmp(p, e) == 0);		// rfc7232,l,548,548
+	}
+
+	/* assume date, strong check [RFC7232 2.2.2 p7] */
+	if (!(ims = VTIM_parse(p)))			// rfc7233,l,502,512
+		return (0);
+
+	/* the response needs a Date */
+	// rfc7232 fc7232,l,439,440
+	if (!http_GetHdr(req->resp, H_Date, &p) || !(d = VTIM_parse(p)))
+		return (0);
+
+	/* grab the Last Modified value */
+	if (http_GetHdr(req->resp, H_Last_Modified, &p)) {
+		lm = VTIM_parse(p);
+		if (!lm)
+			return (0);
+	} else
+		AZ(ObjGetDouble(req->wrk, req->objcore, OA_LASTMODIFIED, &lm));
```

Comment:
Actually, coming back from a de-tour over #3118, I now think that we really should _not_ use `OA_LASTMODIFIED` here:
* While, for the `If-Modified-Since` case, the client could have used some other source to produce the time value for the greater/less than comparison, it can not do this for the `If-Range`: If there was no `Last-Modified` response header, how would it ever have guessed the exact date?

* also I wonder if _Note that this comparison by exact match, including when the validator is an HTTP-date_ from the RFC should be read as "need to be string equal"

So I think now that the modified-date check really should only succeed if `resp.http.Last-Modified` is present.

Reply:
sounds fair, I removed the `OA_LASTMODIFIED` fallback